xarray/conventions.py

```diff
@@ -139,6 +139,15 @@ def _infer_dtype(array, name=None):
     if array.size == 0:
         return np.dtype(float)
 
+    native_dtypes = set(np.vectorize(type, otypes=[object])(array.ravel()))
```

Comment:
since we now look at all objects, not just the first one: could you update the docstring?

Reply:
Done!